feat: Homework2 Created a class for managing animals in zoo

Serhii-Sokolov-McK opened the pull request.

Serhii-Sokolov-McK committed:
feat: Homework2 Created a class for managing animals in zoo

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+
+import "fmt"
+
+// Struct that represents an Animal
+type Animal struct {
```
В загальному, я не прихильник коментарів, де й так дуже зрозумілий код. Але хай вже будуть.

Якщо вже лишати коментарі до структур та полів, то краще робити це по Go-шному. Ось тут детально розписано https://go.dev/doc/comment

Суть в тому, що перше слово має збігатись з назвою структури чи функції. Також наприкінці речення обов'язкова крапка:
```suggestion
// Animal represents an animal in a zoo.
type Animal struct {
```

SerhiiSokolov replied:
Removed all comments

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+type Animal struct {
+	name      string // Name of the animal
+	species   string // Species of the animal
+	inTheCage bool   // Flag indicating whether the animal is in a cage
```
Неймінг - завжди суб'єктивна річ. Але я б перейменував так:
```suggestion
	isCaged bool   // Flag indicating whether the animal is in a cage
```

Зазвичай, але не завжди, `bool` змінні починаються з `is/if` префіксів.

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+// Struct that represents a Zookeeper
+type Zookeeper struct {
```
:memo: Якщо `zookeeper` - одне єдине слово, то ти написав правильно. Якщо два, то правильніше ось так:
```suggestion
type ZooKeeper struct {
```

alexandear replied:
Одне https://www.merriam-webster.com/dictionary/zookeeper

Можна резолвити цей комент.

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+
+// Struct that represents a Zoo, containing multiple animals
+type Zoo struct {
+	Animals []Animal // Slice of animals in the zoo
```
:memo: Ми slices не проходили, не обов'язково було робити.

Щодо коменту. Оскільки `Animals` - експортоване поле, то має бути так (вище лишив лінк на "Go Doc Comments"):

```suggestion
	Animals []Animal // Animals are animals in the zoo.
```

Або так:

```suggestion
	// Animals are animals in the zoo.
	Animals []Animal
```

SerhiiSokolov replied:
Я взагалі повидаляю ці коменти )

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+	// Try to add animals to cages
+	err := zk.AddAnimalToCage(&cages, &zoopark)
+	if err != nil {
+		fmt.Println(err) // Print any error if animals couldn't be placed in cages
```
Я правильно розумію, що якщо наглядач не зможе додати тварину в клітку, то це нормально і програма не має перериватись?

Якщо так, то це треба прокоментувати в коді.

SerhiiSokolov replied:
Додав коменти

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+	return NewAnimal(species, name) // Calls the NewAnimal constructor to create the new animal
+}
+
+// Method for adding animals to cages by the zookeeper
+func (zk Zookeeper) AddAnimalToCage(c *AllCages, z *Zoo) error {
+	// Iterate over the animals in the zoo
+	for i := range z.Animals {
+		animal := &z.Animals[i] // Get a pointer to the current animal
+		if !animal.inTheCage {  // If the animal is not already in a cage
+			// Iterate over the cages
+			for j := range c.Cages {
+				cage := &c.Cages[j]    // Get a pointer to the current cage
+				if cage.animal == "" { // If the cage is empty
+					cage.animal = animal.species // Place the animal in the cage
+					animal.inTheCage = true      // Mark the animal as being in a cage
+					fmt.Printf("Added %s to the cage %s\n", animal.name, cage.name)
+					break // Exit the cage loop as the animal has been placed
+				}
+			}
+			if !animal.inTheCage { // If animal is not in the cage, this means all cages are busy
+				return fmt.Errorf("All cages are occupied, %s is homeless", animal.name) // Return an error
+			}
+		}
+	}
+	return nil // Return nil if all animals were successfully placed in cages
+}
```
Всі ці коментарі зайві, тому що вони дублюють інструкції коду. Можна видаляти.

Краще додати один єдиний коментар до цілої функції. Щось типу:

```go
// AddAnimalToCage садить всіх звірів із зоопарку в клітки.
// Повертає nil, якщо тварина поміщена в клітку.
// Повертає помилку, якщо всі клітки зайняті і тварина не поміщена в клітку.
func (zk Zookeeper) AddAnimalToCage(c *AllCages, z *Zoo) error {
```

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+// Method for adding animals to cages by the zookeeper
+func (zk Zookeeper) AddAnimalToCage(c *AllCages, z *Zoo) error {
+	// Iterate over the animals in the zoo
+	for i := range z.Animals {
+		animal := &z.Animals[i] // Get a pointer to the current animal
+		if !animal.inTheCage {  // If the animal is not already in a cage
```
:memo: Ми поки не вчили `for` та `if`, не буду коментувати їх використання. Можна було не ускладнювати програму :)

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+// Method that describes the current state of the cage
+func (c Cage) DescribeCage() {
+	if c.animal == "" {
+		fmt.Println("Cage is empty!\n") // Print if the cage is empty
```
Ось ця інструкція додасть два переноси рядка. Суфікс `ln` один + `\n` ще один.  

Я думаю, що тут малось на увазі:

```suggestion
		fmt.Println("Cage is empty!")
```

або

```suggestion
		fmt.Printf("Cage is empty!\n")
```

Якщо дійсно треба два переноси рядка, то краще ось так:

```go
		fmt.Printf("Cage is empty!\n\n")
```

alexandear replied:
Коментар `Print if the cage is empty` не потрібен, краще видалити.

SerhiiSokolov replied:
Видалив і комент і \n

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+// Function to create a new animal as a result of reproduction
+func ReproduceAnimals(species string, name string) *Animal {
```
Як думаєш, можливість розмножуватись, це властивість тварини? Тобто чи не краще зробити метод `Reproduce` для `Animal`?

Використання буде по типу:

```go
lion := NewAnimal("lion", "Simba")
kion := lion.Reproduce("Kion")
```

Можна ще далі піти, для розмноження потрібно двоє тварин:

```
lion := NewAnimal("lion", "Simba")
nala := NewAnimal("lion", "Nala")

kion := lion.Reproduce(nala, "Kion") // в льва Simba та львиці Nala народилось дитинча Kion
```

SerhiiSokolov replied:
Додав таку властивість

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+// Struct that represents a Cage, which can hold an animal
+type Cage struct {
+	name   string // Name of the cage
+	animal string // Species of the animal currently in the cage
```
А ми можемо тут замість назви тварини зберігати поінтер на тварину?

```suggestion
	animal *Animal // the animal currently in the cage
```

Поінтер буде `nil`, якщо клітка не зайнята.

SerhiiSokolov replied:
Done

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+// Method that describes the current state of the cage
+func (c Cage) DescribeCage() {
```
Можна без "Cage", так трохи простіше:
```suggestion
func (c Cage) Describe() {
```

alexandear reviewed (changes requested):
Написав кілька зауважень, в цілому хороша робота.

Я би не ускладнював роботу циклами та складними конструкціями, ми це ще не проходили. Але хай вже будуть цикли.

Пропоную видалити більшість коментарів, бо вони не несуть додаткової цінності для читача. Їх підтримка та зміна в майбутньому обійдеться дорожче, ніж просто видалення зараз.

В загальному, коментар в коді залишають тоді, коли не зовсім зрозуміло, для чого цей рядок написаний. Тобто він має відповідати на запитання "Навіщо?" Але бувають і коментарі, які відповідають на питання "Що?" та "Як?" Ось тут деталі https://www.digitalocean.com/community/tutorials/how-to-write-comments-in-go

В умові задачі було "Звіри повтікали". Може додати ще можливість втікати звіру?

Serhii-Sokolov-McK committed:
feat: Homework2
- Removed majority of the comments
- Renamed one field in Animal struct (inTheCage -> isCaged)
- Renamed Cage's method (DescribeCage -> Describe)
-

Serhii-Sokolov-McK committed:
feat: Homework2
- Use pointer to Animal in the cage

Serhii-Sokolov-McK committed:
feat: Homework2
- Added one more lion and cage.
- Made Reproduce method specific to Animals class

Serhii-Sokolov-McK committed:
feat: Homework2
- Animals can escape from cages

SerhiiSokolov commented:
@alexandear Відповів на усі коменти, додав нові методи

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,127 @@
+	}
+	zk := NewZookeeper()
+	err := zk.AddAnimalToCage(&cages, &zoopark)
+	// Якщо усі клітки зайняті то це ок, і доглядач зоопарку має сам вирішити це питання
```
:memo: Коментарі в коді можна писати англійською. Це трішки полегшить тобі роботу, щоб не перемикати розкладки.

SerhiiSokolov replied:
Так ти надав приклад українською, я так і зробив ))

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,127 @@
+
+func NewAnimal(species string, name string) *Animal {
+	return &Animal{species: species, name: name}
+}
+func (a *Animal) Escape() {
+	fmt.Printf("Animal %s escaped from the cage %v\n", a.name, a.cage.name)
+	a.cage.animal = nil
+	a.cage = nil
+}
+func NewCage(name string) *Cage {
```
Між методами, структурами та функціями краще додавати пустий рядок для полегшення читання:

```suggestion
}

func (a *Animal) Escape() {
	fmt.Printf("Animal %s escaped from the cage %v\n", a.name, a.cage.name)
	a.cage.animal = nil
	a.cage = nil
}

func NewCage(name string) *Cage {
```

Інколи методи не розділяються пустими рядками. Але це короткі однорядкові методи, які рідко зустрінеш. Наприклад, https://github.com/golang/go/blob/d36353499f673c89a267a489beb80133a14a75f9/src/cmd/compile/internal/ir/func.go#L188-L192

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,127 @@
+
+type Animal struct {
+	name    string
+	species string
```
Може зробити для `species` константи? У нас же фіксовані значення.

Оголосити тип `type AnimalSpecies string`, який може мати значення:

```
const (
	Lion    AnimalSpecies = "lion"
	Warthog AnimalSpecies = "warthog"
	// other
)
``` 

Змінити всі методи та структури. Замість `species string` буде `species AnimalSpecies`. Це захистить від передачі довільного рядка в `species`.

Serhii-Sokolov-McK replied:
Зробив, але щось я взагалі не зрозумів що це дає бо все одно можу зробити
```go
	simba := NewAnimal("asdkjhasd", "Simba")
	fmt.Println(simba.species)
	nala := NewAnimal(Lion, "Nala")
```

alexandear replied:
Так, чисті raw рядки типу `"asdkjhasd"` можемо передати, тому що цей рядок має тип `untyped string`.

Але якщо у нас є інші константи, то захистить:

```go
type Color string

const (
	Red     Color = "red"
	Yeallow Color = "yellow"
)

nala := NewAnimal(Lion, "Nala")  // скомпілюється
simba := NewAnimal(Red, "Simba") // помилка компіляції `cannot use Red (constant "red" of type Color) as Species value in argument to NewAnimal`
```

Деталі про константи https://go.dev/blog/constants

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,127 @@
+// AddAnimalToCage садить всіх звірів із зоопарку в клітки.
+// Повертає nil, якщо тварина поміщена в клітку.
+// Повертає помилку, якщо всі клітки зайняті і тварина не поміщена в клітку.
+func (zk Zookeeper) AddAnimalToCage(c *AllCages, z *Zoo) error {
```
Пропоную видалити структуру `AllCages`, вона не сильно корисна та ускладнює код. Перенести `Cages []Cage` в структуру `Zoo`. Ніби логічно, що зоопарк містить в собі звірів та клітки. 

Буде так:
```suggestion
func (zk Zookeeper) AddAnimalToCage(z *Zoo) error {
```

Можливо, навіть зоопарк містить в собі наглядача. Але це вже другорядне.

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,127 @@
+	}
+}
+
+func (a Animal) Reproduce(species string, name string) *Animal {
+	return NewAnimal(species, name)
```
Думаю, що `species` не потрібно передавати в параметри. Бо ж ми можемо взяти його із ресівера `a`:
```suggestion
func (a Animal) Reproduce(name string) *Animal {
	return NewAnimal(a.Species, name)
```

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,127 @@
+// Повертає nil, якщо тварина поміщена в клітку.
+// Повертає помилку, якщо всі клітки зайняті і тварина не поміщена в клітку.
+func (zk Zookeeper) AddAnimalToCage(c *AllCages, z *Zoo) error {
+
```
зайвий пустий рядок, краще видалити

alexandear reviewed (changes requested):
Ще кілька зауважень, які пофіксити, і буде чудово.

Serhii-Sokolov-McK committed:
feat: Homework2
- Added const for AnimalSpecies
- Cage is a part of zoo
-

Serhii-Sokolov-McK committed:
feat: Homework2
- Update Reproduce function

alexandear commented on gocourse02/main.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+func main() {
+	simba := NewAnimal("asdkjhasd", "Simba")
+	fmt.Println(simba.species)
```
Повернімо назад, як було:
```suggestion
	simba := NewAnimal(Lion, "Simba")
```

alexandear reviewed (approved):
LGTM

Можеш мерджити після однієї маленької зміни, яку залишив.

Serhii-Sokolov-McK committed:
feat: Homework2
- Some clean up
